Allows to specify a list as seasonal_period and uses multiple seasonalities in TBATS

drmario-gh opened the pull request:
## Related Issuse or bug

First step towards closing #1901.

#### Describe the changes you've made

- Allows to specify a list of values as seasonal_period for a a TimeSeriesExperiment. I introduce a new `all_sp_values` instance variable, so we end up with 3 different instance variables for seasonality values:
  - `self.seasonal_period`: holds the seasonal values specified by the user, as integers.
  - `self.all_sp_values`: holds only the seasonal values that passed the autocorrelation test.
  - `self.sp_to_use`:  always holds a single seasonal value, the first of the `all_sp_values` list.

- Enables specifying multiple seasonalities for TBATS model, by using `all_sp_values` as its `sp` value. The rest of the models are unchanged and keep using `sp_to_use`. 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?
- Slight modifications to `test_seasonal_period_to_use` to assert the new instance variable values.
- Running a TBATS model for airline data with [12, 6] seasonality. 

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)

`NA`

## Checklist:
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

ngupta23 commented on pycaret/internal/pycaret_experiment/time_series_experiment.py:
```diff
@@ -139,6 +140,7 @@ def _get_setup_display(self, **kwargs) -> Styler:
                 ["Seasonal Period Tested", self.seasonal_period],
                 ["Seasonality Detected", self.seasonality_present],
                 ["Seasonality Used in Models", self.sp_to_use],
+                ["Seasonalities Used in Multiple Seasonality Models", self.all_sp_values],
```
If you do not mind, can you rename this to the following:
```python
["Seasonal Period(s) Tested", self.seasonal_period],
["Seasonality Present", self.seasonality_present],
["Seasonalities Detected", self.all_sp_values],
["Primary Seasonality", self.sp_to_use],
```

drmario-gh replied:
Sure!

ngupta23 commented on pycaret/internal/pycaret_experiment/time_series_experiment.py:
```diff
@@ -364,7 +366,7 @@ def _check_and_clean_index(
             then the data's index is used as is for modeling.
 
-        seasonal_period : Optional[Union[int, str]], default = None
+        seasonal_period : Optional[Union[List[Union[int, str]], int, str]], default = None
```
Now that this is more complex, it may be good to add some examples here for the different use cases. What do you think?

ngupta23 commented on pycaret/internal/pycaret_experiment/time_series_experiment.py:
```diff
@@ -484,56 +486,86 @@ def _check_and_set_fh(
+
+        Parameters
+        ----------
+        seasonal_period : Optional[Union[List[Union[int, str]], int, str]]
```
In this method, it will only be of type int or str (and not List) - correct?

drmario-gh replied:
Oops, sorry! You are right. I got confused when reviewing the type hints.

ngupta23 commented on pycaret/tests/test_time_series_setup.py:
```diff
@@ -154,13 +154,24 @@ def test_seasonal_period_to_use():
         data=data, fh=fh, verbose=False, session_id=42,
     )
     assert exp.sp_to_use == 12
+    assert exp.all_sp_values == [12]
```
Maybe add a test for `self.seasonal_period` just to be on the safe side?

ngupta23 commented on pycaret/tests/test_time_series_setup.py:
```diff
@@ -154,13 +154,24 @@ def test_seasonal_period_to_use():
+        data=data, fh=fh, verbose=False, session_id=42, seasonal_period=['M', 6]
+    )
+    assert exp.all_sp_values == [12, 6]
+    assert exp.sp_to_use == 12
```
Maybe add a test for self.seasonal_period just to be on the safe side?

ngupta23 commented on pycaret/tests/test_time_series_setup.py:
```diff
@@ -154,13 +154,24 @@ def test_seasonal_period_to_use():
+    
     # Should get 1 even though we passed 12
+    assert exp.all_sp_values == [1]
     assert exp.sp_to_use == 1
```
Maybe add a test for self.seasonal_period just to be on the safe side?

ngupta23 commented:
@drmario-gh Thank you again for your PR. Your changes look good and are well structured. I left some minor comments in there that should hopefully be quick to resolve. Thanks again!

drmario-gh commented on pycaret/internal/pycaret_experiment/time_series_experiment.py:
```diff
@@ -484,56 +486,86 @@ def _check_and_set_fh(
+            if seasonality_present
+        ] or [1]
+        self.sp_to_use = self.all_sp_values[0]
+        self.seasonal_period = seasonal_period[0]
```
Makes sense. I wasn't sure at that moment if `seasonal_period` was in use somewhere else and thus, I kept it as a single value. I have double checked now and yes, it seems we can do so.

Just in case, I still turn it into an scalar value if the `seasonal_period` list has only one element for retrocompatibility with the users (e.g., somebody using `exp.seasonal_period`  and suddenly receiving a list of one element instead of a value).

drmario-gh commented:
> @drmario-gh Thank you again for your PR. Your changes look good and are well structured. I left some minor comments in there that should hopefully be quick to resolve. Thanks again!

Thank you, @ngupta23! I think I have addressed all your comments :)
